### ISSUE TITLE
non-constexpr formatted_size() breaks compile-time formatting

### ISSUE THREAD

**Issue by AnthonyVH** [fmtlib/fmt#4102]
Before [f6b4a23b83](https://github.com/fmtlib/fmt/commit/f6b4a23b83b36946072398bd04b7c7810df2d213) the function `formatted_size()` in `compile.h` was marked as `FMT_CONSTEXPR20`.

This allowed writing compile-time formatting using C++23's `static constexpr` members inside a constexpr function. E.g.:

```c++
template <fmt::detail::udl_compiled_string format, auto... values>
constexpr std::string_view constexpr_format() {
  static constexpr auto str = [] {
    constexpr auto result_length = fmt::formatted_size(format, values...);
    auto result = std::array<char, result_length>{};
    fmt::format_to(result.data(), format, values...);
    return result;
  }();
  return std::string_view(str.data(), str.size());
}
```

However, since the linked commit, this annotation was removed. Looking at the commit, I can't figure out why this is the case. Is there any chance of re-introducing the `constexpr` on all functions that allow it in `compile.h`?

PS: `fmt::counting_buffer::count()` is not marked `constexpr` either. When I wrote the snippet above, an `fmt::counting_iterator` was used instead in `fmt::formatted_size()` (but this iterator was removed in [ba36a0481](https://github.com/fmtlib/fmt/commit/ba36a04811c5f73681a8e1c1e28539182e6a211b)).

**Comment by vitaut** [comment]
It was relying on deprecated and inefficient `counting_iterator` which was removed. A PR to constexprify `counting_buffer` and reintroduce `constexpr` to `formatted_size` would be welcome.

**Comment by phprus** [comment]
Fix: #4103

**Comment by AnthonyVH** [comment]
I was just about to prepare a PR, but you beat me to it :smile:. Thanks a lot for the super quick fix!
